misc/python/materialize/output_consistency/ignore_filter/inconsistency_ignore_filter.py

```diff
@@ -0,0 +1,96 @@
+from materialize.output_consistency.expression.expression_characteristics import (
+    ExpressionCharacteristics,
+)
+from materialize.output_consistency.expression.expression_with_args import (
+    ExpressionWithArgs,
+)
+from materialize.output_consistency.expression.leaf_expression import LeafExpression
+from materialize.output_consistency.operation.operation import DbFunction
+from materialize.output_consistency.selection.selection import DataRowSelection
+
+
+class InconsistencyIgnoreFilter:
+    """Allows specifying and excluding expressions with known output inconsistencies"""
+
+    def shall_ignore(
+        self, expression: Expression, row_selection: DataRowSelection
+    ) -> bool:
+        if isinstance(expression, LeafExpression):
+            return False
+        elif isinstance(expression, ExpressionWithArgs):
+            return self._shall_ignore_expression_with_args(expression, row_selection)
+        else:
+            raise RuntimeError(f"Unsupported expression type: {type(expression)}")
+
+    def _shall_ignore_expression_with_args(
+        self,
+        expression: ExpressionWithArgs,
+        row_selection: DataRowSelection,
+    ) -> bool:
+        # check expression itself
+        if self._shall_ignore_expression_with_args_no_recursion(
+            expression, row_selection
+        ):
+            return True
+
+        # recursively check arguments
+        for arg in expression.args:
+            if self.shall_ignore(arg, row_selection):
+                return True
+
+        return False
+
+    def _shall_ignore_expression_with_args_no_recursion(
+        self,
+        expression: ExpressionWithArgs,
+        row_selection: DataRowSelection,
+    ) -> bool:
+        if not expression.operation.is_aggregation:
+            # currently no issues without aggregation are known
+            return False
+
+        if isinstance(expression.operation, DbFunction):
+            # currently only issues with functions are known, therefore ignore operations
+            db_function = expression.operation
+            expression_characteristics = (
+                expression.recursively_collect_involved_characteristics(row_selection)
+            )
+
+            if self._matches_problematic_function_invocation(
+                db_function, expression_characteristics
+            ):
+                return True
+
+        return False
+
+    def _matches_problematic_function_invocation(
+        self,
+        db_function: DbFunction,
+        all_involved_characteristics: Set[ExpressionCharacteristics],
+    ) -> bool:
+        # Note that function names are always provided in lower case.
+        if db_function.function_name in {"sum", "avg", "stddev_samp", "stddev_pop"}:
+            if ExpressionCharacteristics.MAX_VALUE in all_involved_characteristics:
+                # tracked with https://github.com/MaterializeInc/materialize/issues/19511
```

Comment:
Do we actually still generate those queries or just not generate them at all then? If we do, could be interesting to keep track of the number of queries failing with this issue.

Reply:
Filtering happens after generation, we can count them. Will do.

Reply:
Not addressed for now (see previous comment), let me know if this is important to you.